### PR TITLE
Issue #7594  - Charts widget colour coded classification

### PR DESCRIPTION
**Pull request by ale-cristofori**
## Description
This PR designs a new functionality, allowing users to set their own colour coded attribute classifications and labels for bar and pie chart widgets. The feature was requested for a downstream MS2 project

[#C070#214](https://github.com/geosolutions-it/npa-cgg/issues/214)
[#C070#215](https://github.com/geosolutions-it/npa-cgg/issues/215)

This PR incorporates some of the changes and further expand some previous work that was done for the issue above, the PR for that work is https://github.com/geosolutions-it/MapStore2/pull/7484

This feature has got to be implemented in MS2 master, in C070 downstream project master and ported to 2021.01.xx

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [ ] Bugfix
 - [x] Feature
 - [ ] Code style update (formatting, local variables)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build related changes
 - [ ] CI related changes
 - [ ] Other... Please describe:

## Issue

**What is the current behavior?**
From the catalog load a layer on the that supports WFS and WPS services.
Open the TOC and select the widget button (bar chart icon on the toolbar).
On the Widget TOC dialog select Chart and subsequently Bar Chart or Pie Chart.

On the Chart wizard (either Bar or Pie):
For bar chart: select the `X` attribute , the `Y` attribute and `operation` (choice could also be `no operation`). Select a color from the `Color` drop down
For pie chart:  select the `group By`, `Use` attribute and `operation` (choice could also be `no operation`). Select a color ramp from the `Color Ramp` drop down.
toggle on the `Display Legend` toggle.

The respective charts will be coloured according to the colours chosen, colours for the chart items won't follow a classification on any additional attribute.

[#7594](https://github.com/geosolutions-it/MapStore2/issues/7594)

**What is the new behavior?**
From the catalog load a layer on the that supports WFS and WPS services.
Open the TOC and select the widget button (bar chart icon on the toolbar).
On the Widget TOC dialog select Chart and subsequently Bar Chart or Pie Chart.

On the Chart wizard (either Bar or Pie):
For bar chart: select the `X` attribute , the `Y` attribute and `operation` (choice could also be `no operation`). Select the `Custom` option from the `Color` drop down, last item in the list.
For pie chart:  select the `group By`, `Use` attribute and `operation` (choice could also be `no operation`). Select the `Custom` option from the `Color Ramp` drop down, last item in the list.
toggle on the `Display Legend` toggle.

A modal window will open with:
1. A colour picker to choose a new default colour (instead of the pre-set ones, for Pie chart this will colour all slices evenly, there will be no colour ramp at this time.
2. A `Classification Attribute` drop down, allowing the user to select the attribute they want the colour coded classification to be based upon. The selection of one of the attributes (those will be the ones in the TOC selected layer) from the drop down list will open.
3. A text box `Default Class Label` that will allow the user to define the text for the `Default` label value (by default is `Default`)
4. A table row with three columns `Class Color` `Class Value` and `Class Label`.

Users can now define their own colour, with a colour picker that will pop up when they click on the colour swatch, the value to colour code the classification (this must exactly match the value on the attribute layer) and a custom label for the legend for the attribute (if no custom label is set the value will be used).

Users can add/remove rows to the table to define further classifications, clicking on the menu on the right (the three dots), it won't be possible to remove all classifications at least one must be there. Users can clear the colour coded classification and revert to non-colour coded one clearing the selection on the `Classification Attribute` drop down list.

When the users finish their colour coded selections the can click on the `Confirm` button to apply the desired colour coded classification, the chart will be styled accordingly.

When Users click on the `Cancel` button, if a colour coded classification is defined it will be discarded and the chart colouring will fall back to the default colour with no classification.

On the widget, when toggling on the legend observe the legend items grouped by user defined classification with custom labels if defined for bar chart and custom label values near every single item for pie chart.  

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

## Other useful information
recording_1

### CODE REVIEW

**Review by offtherailz — CHANGES_REQUESTED** [review]
Quite impressive :rocket: 

Only a couple of things: 
- [ ] The legend for pie should be grouped, as well as for bar charts.
screenshot-localhost_8081-2021 12 06-18_21_32
- [ ] Custom entry, when selected, should show a button on the side to open configuration window (without selecting again "Custom" entry, that is not intuitive). Handle it outside the color ramp selector, if possible. 
- [ ] "Custom" entry should be colored, if possible, with the colors selected.

**Comment by offtherailz** [comment]
Ok, please investigate a little on that, but if you find is not feasible, you can stop. I'll review this in the meanwhile. Thank you.

**Comment by offtherailz** [comment]
Ah, please. Fix the unit tests

**Review by offtherailz — CHANGES_REQUESTED** [review]
Good. I like it a lot. I've only some changes to apply: 


- [x] If I hit `cancel` the customization is partially cleared (default color keept, other colors removed). This is not what I expect from a "cancel" button, but more from a "clear" button. Because I do several customization in that window, I'd prefer to not having the cancel button at all. So please remove it.
- [x] Pie chart
   - [x] Title of the legend should not be shown, as for the original pie chart (we can allow sooner or later to insert it)
   - [x]  Do not write the value in the pie section (You can not control the size of these string, so they can become very big or not readable. We can allow it in the future, as an option, but now please disable).
- [ ] Legends: 
   - [ ] Legend text is a little ugly sometimes.
image. The main problem is related to the fixed `<label1> - <label2>` format. When the strings are empty, you may have a `-` appended. 

Moreover you can have charts when the classification attribute is the same of the group-by attribute. 
image In this case the first value before the '-' is a repetition, we should allow users to completely override the entire label. A good idea good idea would be using `${}` as replacement. For lines you will have a similar issue with `count(ATTR) - test` where `count(ATTR)` is redundant. I suggest you change this behaviour, setting a custom label a template. `${legendValue}` can be used as a placeholder for the legend value label, that can be the attribute name (pie) or the legend label (chart). In the legend you will simply replace it with string `.replace()` function. 

Examples: 

- Pie: 
   - "`${legendValue}` (Mountain)" --> Arizona (Montain)
   - "Mountain" --> Mountain"
 - Bar: 
    - "`${legendValue}` - Mountain" --> SUM(PERSONS) - Mountain
         - *Setting legend label="persons" --> persons - Mountain
- [ ] Add an hint to suggest these how to use the placeholder in the color customization window

File: web/client/configs/localConfig.json [hunk 1]
```diff
@@ -41,12 +41,7 @@
       }, {
         "urlPattern": ".*rest/config.*",
         "method": "bearer"
-      },
-      {
-        "urlPattern": "http(s)?\\:\\/\\/gs-stable\\.geo-solutions\\.it(\\:443|\\:80)?\\/geoserver/.*",
-        "authkeyParamName": "authkey",
-        "method": "authkey"
-    }],
+        }],
```
Comment (offtherailz) [inline web/client/configs/localConfig.json:44]:
```suggestion
         },
      {
        "urlPattern": "http(s)?\\:\\/\\/gs-stable\\.geo-solutions\\.it(\\:443|\\:80)?\\/geoserver/.*",
        "authkeyParamName": "authkey",
        "method": "authkey"
    }],
```

File: web/client/configs/localConfig.json [hunk 2]
```diff
@@ -41,7 +41,12 @@
       }, {
         "urlPattern": ".*rest/config.*",
         "method": "bearer"
-        }],
+      },
+      {
+        "urlPattern": "http(s)?\\:\\/\\/gs-stable\\.geo-solutions\\.it(\\:443|\\:80)?\\/geoserver/.*",
+        "authkeyParamName": "authkey",
+        "method": "authkey"
+    }],
```
Comment (offtherailz) [inline web/client/configs/localConfig.json:49]:
```suggestion
        }],
```

File: web/client/translations/data.en-US.json [hunk 3]
```diff
@@ -1781,9 +1781,12 @@
                         "classValue": "Class Value",
                         "color": "Color",
                         "classificationAttribute": "Classification Attribute",
+                        "customLabels": "Custom Labels",                        
+                        "customLabelsExample": "<div><p><code>${legendValue}</code> can be used as placeholder to replace the legend original value.</p><p>legend items with the same text content will be grouped in the chart.</p><h5>Example</h5><dl><dt><b>Attribute:</b></dt><dd>POPULATION</dd><dt><b>Class Label:</b></dt><dd><code>${legendValue}</code> California</dd><dt><b>Result:</b></dt><dd>POPULATION California</dd></dl></div>",
                         "defaultClassLabel": "Default Class Label",
                         "defaultColorRamp": "Default Color Ramp",
                         "defaultColor": "Default Color",
+                        "defaultCustomLabel": "Default Custom Label",
```
Comment (offtherailz) [inline web/client/translations/data.en-US.json:1789]:
```suggestion
                        "defaultCustomLabel": "Default Label",
```

File: web/client/translations/data.es-ES.json [hunk 4]
```diff
@@ -1780,9 +1780,12 @@
                         "classValue": "Valor de la Clase",
                         "color": "Color",
                         "classificationAttribute": "Atributo de clasificación",
+                        "customLabels": "Etiquetas personalizadas",
+                        "customLabelsExample": "<div><p><code>${legendValue}</code> se puede utilizar como marcador de posición para reemplazar el valor de la etiqueta original.</p><p>Los elementos de leyenda con el mismo contenido de texto se agrupan en el diagrama.</p><h5>Ejemplo</h5><dl><dt><b>Atributo:</b></dt><dd>POPULATION</dd><dt><b>Etiqueta de la Clase:</b></dt><dd><code>${legendValue}</code> California</dd><dt><b>Resultado:</b></dt><dd>POPULATION California</dd></dl></div>",
                         "defaultClassLabel": "Etiqueta de Clase Predeterminada",
                         "defaultColorRamp": "Rampa de Color Predeterminada",
                         "defaultColor": "Color Predeterminado",
+                        "defaultCustomLabel": "Etiqueta predeterminada personalizada",
```
Comment (offtherailz) [inline web/client/translations/data.es-ES.json:1788]:
```suggestion
                        "defaultCustomLabel": "Etiqueta predeterminada",
```

File: web/client/translations/data.fr-FR.json [hunk 5]
```diff
@@ -1781,9 +1781,12 @@
                         "classValue": "Valeur de la Classe",
                         "color": "Couleur",
                         "classificationAttribute": "Attribut de Classification",
+                        "customLabels": "Étiquettes personnalisées",
+                        "customLabelsExample": "<div><p><code>${legendValue}</code> peut être utilisé comme espace réservé pour remplacer la valeur de la légende d'origine.</p><p>Les éléments de légende avec le même contenu textuel seront regroupés dans le graphique.</p><h5>Exemple</h5><dl><dt><b>Attribut:</b></dt><dd>POPULATION</dd><dt><b>Étiquette de Classe:</b></dt><dd><code>${legendValue}</code> California</dd><dt><b>Résultat:</b></dt><dd>POPULATION California</dd></dl></div>",
                         "defaultClassLabel": "Étiquette de Classe par Défaut",
                         "defaultColorRamp": "Rampe de Couleur par Défaut",
                         "defaultColor": "Couleur par Défaut",
+                        "defaultCustomLabel": "Étiquette par défaut personnalisée",
```
Comment (offtherailz) [inline web/client/translations/data.fr-FR.json:1789]:
```suggestion
                        "defaultCustomLabel": "Étiquette par défaut",
```

File: web/client/translations/data.it-IT.json [hunk 6]
```diff
@@ -1780,9 +1780,12 @@
                         "classValue": "Valore della Classe",
                         "color": "Colore",
                         "classificationAttribute": "Attributo di Classificazione",
+                        "customLabels": "Etichette personalizzate",
+                        "customLabelsExample": "<div><p><code>${legendValue}</code> può essere utilizzato come segnaposto per sostituire il valore originale della legenda.</p><p>Gli elementi della legenda con lo stesso contenuto di testo verranno raggruppati nel grafico.</p><h5>Esempio</h5><dl><dt><b>Attributo:</b></dt><dd>POPULATION</dd><dt><b>Etichetta della Classe:</b></dt><dd><code>${legendValue}</code> California</dd><dt><b>Risultato:</b></dt><dd>POPULATION California</dd></dl></div>",
                         "defaultClassLabel": "Etichetta della Classe di Default",
                         "defaultColorRamp": "Rampa di Colori di Default",
                         "defaultColor": "Colore di Default",
+                        "defaultCustomLabel": "Etichetta di Default personalizzata",
```
Comment (offtherailz) [inline web/client/translations/data.it-IT.json:1788]:
```suggestion
                        "defaultCustomLabel": "Etichetta di Default",
```

File: web/client/translations/data.de-DE.json [hunk 7]
```diff
@@ -1811,7 +1811,22 @@
...
+                        "classLabel": "Klassenbezeichnung",
+                        "classValue": "Klassenwert",
+                        "color": "Farbe",
+                        "classificationAttribute": "Klassifizierungsattribut",
+                        "customLabels": "Benutzerdefinierte Etiketten",
+                        "customLabelsExample": "<div><p><code>${legendValue}</code> kann als Platzhalter verwendet werden, um den Wert der ursprünglichen Beschriftung zu ersetzen.</p><p>Legendenelemente mit gleichem Textinhalt werden im Diagramm gruppiert.</p><h5>Beispiel</h5><dl><dt><b>Attribut:</b></dt><dd>POPULATION</dd><dt><b>Klassenbezeichnung:</b></dt><dd><code>${legendValue}</code> California</dd><dt><b>Ergebnisse:</b></dt><dd>POPULATION California</dd></dl></div>",
+                        "defaultClassLabel": "Standardklassenbezeichnung",
+                        "defaultColorRamp": "Standard-Farbrampe",
+                        "defaultColor": "Standardfarbe",
+                        "defaultCustomLabel": "Benutzerdefiniertes Standardlabel",
```
Comment (offtherailz) [inline web/client/translations/data.de-DE.json:1826]:
```suggestion
                        "defaultCustomLabel": "Standardlabel",
```

**Review by offtherailz — APPROVED** [review]

**Comment by ElenaGallo** [comment]
@ale-cristofori after my test I only notice a small improvement to make if possible:

there should be the option to delete the first value in the Class Color on the *Custom Colors Settings* panel.

*Current Result*
The Delete option is not present for the first value


*Expected Result*
The Delete option is present for the first value as is the case for subsequent values

**Comment by tdipisa** [comment]
@ElenaGallo please put this comment in the related issue

**Comment by ale-cristofori** [comment]
> @ale-cristofori after my test I only notice a small improvement to make if possible:
> 
> there should be the option to delete the first value in the Class Color on the _Custom Colors Settings_ panel.
> 
> _Current Result_ The Delete option is not present for the first value 
> 
> _Expected Result_ The Delete option is present for the first value as is the case for subsequent values 

Hi @ElenaGallo , that is intentional, to avoid users removing all items from the form, thus removing chart styles. It could be made more 'clever,  so that the the first line cannot be removed if it is the only one ( the remanining one)

**Comment by tdipisa** [comment]
> Hi @ElenaGallo , that is intentional, to avoid users removing all items from the form, thus removing chart styles. It could be made more 'clever, so that the the first line cannot be removed if it is the only one ( the remanining one)

yes indeed

**Pull request merged**